planning/mission_planner/src/mission_planner/route_selector.cpp

```diff
@@ -0,0 +1,234 @@
+  sub_route_ = create_subscription<LaneletRoute>(
+    "~/planner/route", durable_qos, std::bind(&RouteSelector::on_route, this, _1));
+
+  // Set initial state.
+  const auto stamp = now();
+  main_.change_state(stamp, RouteState::INITIALIZING);
+  mrm_.change_state(stamp, RouteState::INITIALIZING);
+  initialized_ = false;
+  mrm_operating_ = false;
+  main_request_ = std::monostate{};
+}
+
+void RouteSelector::on_state(const RouteState::ConstSharedPtr msg)
+{
+  if (msg->state == RouteState::UNSET && !initialized_) {
+    const auto stamp = now();
+    main_.change_state(stamp, RouteState::UNSET);
+    mrm_.change_state(stamp, RouteState::UNSET);
+    initialized_ = true;
+  }
+
+  (mrm_operating_ ? mrm_ : main_).update_state(*msg);
+}
+
+void RouteSelector::on_route(const LaneletRoute::ConstSharedPtr msg)
+{
+  (mrm_operating_ ? mrm_ : main_).update_route(*msg);
+}
+
+void RouteSelector::on_clear_route_main(
+  ClearRoute::Request::SharedPtr req, ClearRoute::Response::SharedPtr res)
+{
+  // Save the request to resume from MRM.
+  main_request_ = std::monostate{};
+
+  // During MRM, only change the state.
+  if (mrm_operating_) {
+    main_.change_state(now(), RouteState::UNSET);
+    res->status.success = true;
+    return;
+  }
+
+  // Forward the request if not in MRM.
+  res->status = service_utils::sync_call(cli_clear_route_, req);
+}
+
+void RouteSelector::on_set_waypoint_route_main(
+  SetWaypointRoute::Request::SharedPtr req, SetWaypointRoute::Response::SharedPtr res)
+{
+  // Save the request to resume from MRM.
+  req->uuid = uuid::generate_if_empty(req->uuid);
+  main_request_ = req;
+
+  // During MRM, only change the state.
+  if (mrm_operating_) {
+    main_.change_state(now(), RouteState::SET);
+    res->status.success = true;
+    return;
+  }
+
+  // Forward the request if not in MRM.
+  res->status = service_utils::sync_call(cli_set_waypoint_route_, req);
+}
+
+void RouteSelector::on_set_lanelet_route_main(
+  SetLaneletRoute::Request::SharedPtr req, SetLaneletRoute::Response::SharedPtr res)
+{
+  // Save the request to resume from MRM.
+  req->uuid = uuid::generate_if_empty(req->uuid);
+  main_request_ = req;
+
+  // During MRM, only change the state.
+  if (mrm_operating_) {
+    main_.change_state(now(), RouteState::SET);
+    res->status.success = true;
+    return;
+  }
+
+  // Forward the request if not in MRM.
+  res->status = service_utils::sync_call(cli_set_lanelet_route_, req);
+}
+
+void RouteSelector::on_clear_route_mrm(
+  ClearRoute::Request::SharedPtr req, ClearRoute::Response::SharedPtr res)
+{
+  const auto create_resume_request = [](const auto request) {
+    const auto r = std::make_shared<SetWaypointRoute::Request>();
+    r->header = request->header;
+    r->goal_pose = request->goal_pose;
+    r->uuid = request->uuid;
+    r->allow_modification = request->allow_modification;
+    return r;
+  };
+
+  mrm_operating_ = false;
+  mrm_.change_state(now(), RouteState::UNSET);
```

Comment:
clear route mrm sometimes fails, so better to set this after reroute wit normal route success


in the previous code, keep mrm route when reoute fails
https://github.com/autowarefoundation/autoware.universe/blob/main/planning/mission_planner/src/mission_planner/mission_planner.cpp#L525-L529

Reply:
Fix to clear mrm mode if the route clear/change is succeeded. https://github.com/autowarefoundation/autoware.universe/pull/6363/commits/4951e1b054dd8a3b29ff2cccf89e02305cff2b22#diff-b8e18a7cd1279c2c60737e4e006028710a1b6f18b055a63d1070c0f3f8a659c1R194-R197.